File: tiktok_live_state_assembler.py
"""
Live state assembly layer for TikTok SaaS frontend consumption.

Consolidates all backend modules into one deterministic UI-ready object.
This does NOT merge intelligence — it only normalises outputs with safe defaults.
Never raises; never returns null; never omits contract fields.
"""

from __future__ import annotations

import importlib
import logging
from typing import Any, Callable

logger = logging.getLogger(__name__)

_SOURCE_MODULES: tuple[tuple[str, str], ...] = (
    ("trend_detection_engine", "get_state"),
    ("emerging_products_engine", "get_state"),
    ("trending_products_engine", "get_state"),
    ("inventory_prevention_system", "get_state"),
    ("inventory_gap_system", "get_state"),
    ("content_queue_system", "get_state"),
    ("approval_system", "get_state"),
    ("performance_tracker", "get_state"),
    ("learning_engine", "get_state"),
    ("system_health_monitor", "get_state"),
)


def _empty_contract() -> dict[str, Any]:
    return {
        "today_flow": {
            "step": "trend → product → content → queue",
            "next_action": "unknown",
            "status": "unknown",
        },
        "trends": [],
        "products": [],
        "inventory_gaps": [],
        "inventory_prevention": [],
        "content_queue": [],
        "approvals": [],
        "performance": {},
        "alerts": [],
        "primary_action": {
            "label": "unknown",
            "action": "unknown",
            "context_id": "unknown",
        },
        "system_health": "unknown",
    }


def _as_list(value: Any) -> list:
    return value if isinstance(value, list) else []


def _as_dict(value: Any) -> dict:
    return value if isinstance(value, dict) else {}


def _as_string(value: Any, default: str = "unknown") -> str:
    if value is None:
        return default
    text = str(value).strip()
    return text if text else default


def _as_number(value: Any, default: float = 0) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return default


def _safe_invoke(module_name: str, fn_name: str, account_id: str) -> dict[str, Any]:
    try:
        module = importlib.import_module(module_name)
        fn: Callable[[str], Any] | None = getattr(module, fn_name, None)
        if not callable(fn):
            return {}
        result = fn(account_id)
        return result if isinstance(result, dict) else {}
    except Exception as exc:
        logger.warning("assembleLiveState: %s.%s failed: %s", module_name, fn_name, exc)
        return {}


def _merge_products(emerging: list, trending: list) -> list:
    merged: list[dict[str, Any]] = []
    seen: set[str] = set()
    for item in emerging + trending:
        if not isinstance(item, dict):
            continue
        name = _as_string(item.get("name") or item.get("product_name"), "")
        key = name.lower()
        if not key or key in seen:
            continue
        seen.add(key)
        merged.append({
            "name": _as_string(item.get("name") or item.get("product_name")),
            "signal_strength": _as_number(item.get("signal_strength") or item.get("score")),
            "source": _as_string(item.get("source")),
            "confidence": _as_number(item.get("confidence")),
            "evidence": _as_list(item.get("evidence")),
        })
    return merged


def _build_alerts(
    inventory_gaps: list,
    inventory_prevention: list,
    system_health: str,
    partial_failures: list[str],
) -> list[dict[str, str]]:
    alerts: list[dict[str, str]] = []

    for gap in inventory_gaps:
        if not isinstance(gap, dict):
            continue
        alerts.append({
            "level": "warning",
            "code": "inventory_gap",
            "message": f"Inventory gap: {_as_string(gap.get('product_name'))}",
        })

    for item in inventory_prevention:
        if not isinstance(item, dict):
            continue
        if item.get("available") is False:
            alerts.append({
                "level": "warning",
                "code": "inventory_prevention",
                "message": f"Prevent stockout: {_as_string(item.get('product_name'))}",
            })

    if system_health == "failing":
        alerts.append({
            "level": "error",
            "code": "system_health",
            "message": "System health is failing",
        })
    elif system_health == "degraded":
        alerts.append({
            "level": "warning",
            "code": "system_health",
            "message": "System health is degraded",
        })

    for module_name in partial_failures:
        alerts.append({
            "level": "info",
            "code": "module_fallback",
            "message": f"Using defaults for {module_name}",
        })

    return alerts


def _derive_flow_and_action(
    trends: list,
    products: list,
    content_queue: list,
    approvals: list,
    inventory_gaps: list,
    system_health: str,
) -> tuple[dict[str, str], dict[str, str]]:
    today_flow = {
        "step": "trend → product → content → queue",
        "next_action": "unknown",
        "status": "unknown",
    }
    primary_action = {
        "label": "unknown",
        "action": "unknown",
        "context_id": "unknown",
    }

    if inventory_gaps:
        today_flow["next_action"] = "Resolve inventory gaps before attaching products"
        today_flow["status"] = "blocked"
        gap = inventory_gaps[0] if isinstance(inventory_gaps[0], dict) else {}
        primary_action = {
            "label": "Fix inventory gap",
            "action": "resolve_inventory_gap",
            "context_id": _as_string(gap.get("product_name")),
        }
        return today_flow, primary_action

    if approvals:
        today_flow["next_action"] = "Review pending content approvals"
        today_flow["status"] = "awaiting_approval"
        item = approvals[0] if isinstance(approvals[0], dict) else {}
        primary_action = {
            "label": "Approve content",
            "action": "approve_content",
            "context_id": _as_string(item.get("content_id")),
        }
        return today_flow, primary_action

    if not content_queue and products:
        today_flow["next_action"] = "Generate content from detected products"
        today_flow["status"] = "ready_for_content"
        product = products[0] if isinstance(products[0], dict) else {}
        primary_action = {
            "label": "Generate content",
            "action": "generate_content",
            "context_id": _as_string(product.get("name")),
        }
        return today_flow, primary_action

    if not products and trends:
        today_flow["next_action"] = "Match products to active trends"
        today_flow["status"] = "trend_detected"
        trend = trends[0] if isinstance(trends[0], dict) else {}
        primary_action = {
            "label": "Match products",
            "action": "match_products",
            "context_id": _as_string(trend.get("id") or trend.get("summary")),
        }
        return today_flow, primary_action

    if content_queue:
        today_flow["next_action"] = "Monitor queued content pipeline"
        today_flow["status"] = "in_queue"
        item = content_queue[0] if isinstance(content_queue[0], dict) else {}
        primary_action = {
            "label": "View queue",
            "action": "view_queue",
            "context_id": _as_string(item.get("id")),
        }
        return today_flow, primary_action

    if system_health in ("healthy", "degraded", "failing"):
        today_flow["status"] = system_health
        today_flow["next_action"] = "Run trend scan to refresh signals"
        primary_action = {
            "label": "Refresh trends",
            "action": "run_trend_scan",
            "context_id": "unknown",
        }

    return today_flow, primary_action


def assembleLiveState(account_id: str) -> dict[str, Any]:
    """
    Assemble the strict UI contract from all optional backend modules.

    Never raises. Never returns null. Every field is always present.
    """
    state = _empty_contract()
    partial_failures: list[str] = []

    collected: dict[str, Any] = {}
    for module_name, fn_name in _SOURCE_MODULES:
        payload = _safe_invoke(module_name, fn_name, account_id)
        if not payload and module_name not in collected:
            partial_failures.append(module_name)
        collected[module_name] = payload

    trends = _as_list(collected.get("trend_detection_engine", {}).get("trends"))
    emerging = _as_list(collected.get("emerging_products_engine", {}).get("products"))
    trending = _as_list(collected.get("trending_products_engine", {}).get("products"))
    products = _merge_products(emerging, trending)

    inventory_gaps = _as_list(collected.get("inventory_gap_system", {}).get("inventory_gaps"))
    inventory_prevention = _as_list(
        collected.get("inventory_prevention_system", {}).get("inventory_prevention")
    )
    content_queue = _as_list(collected.get("content_queue_system", {}).get("content_queue"))
    approvals = _as_list(collected.get("approval_system", {}).get("approvals"))
    performance = _as_dict(collected.get("performance_tracker", {}).get("performance"))
    system_health = _as_string(
        collected.get("system_health_monitor", {}).get("system_health"),
        "unknown",
    )

    today_flow, primary_action = _derive_flow_and_action(
        trends=trends,
        products=products,
        content_queue=content_queue,
        approvals=approvals,
        inventory_gaps=inventory_gaps,
        system_health=system_health,
    )

    alerts = _build_alerts(
        inventory_gaps=inventory_gaps,
        inventory_prevention=inventory_prevention,
        system_health=system_health,
        partial_failures=partial_failures,
    )

    state["today_flow"] = today_flow
    state["trends"] = trends
    state["products"] = products
    state["inventory_gaps"] = inventory_gaps
    state["inventory_prevention"] = inventory_prevention
    state["content_queue"] = content_queue
    state["approvals"] = approvals
    state["performance"] = performance
    state["alerts"] = alerts
    state["primary_action"] = primary_action
    state["system_health"] = system_health

    return state

File: api/tiktok-live-state-assembler.js
/**
 * Live state assembler (Node) — mirrors tiktok_live_state_assembler.py contract.
 * Safe UI contract builder only; never throws.
 */

const DEFAULT_FEED_TABLE = "trend_intelligence_feed";
const DEFAULT_QUEUE_TABLE = "content_queue";
const DEFAULT_PERFORMANCE_TABLE = "content_performance";
const DEFAULT_CACHE_TABLE = "tiktok_insights_cache";

function emptyContract() {
  return {
    today_flow: {
      step: "trend → product → content → queue",
      next_action: "unknown",
      status: "unknown",
    },
    trends: [],
    products: [],
    inventory_gaps: [],
    inventory_prevention: [],
    content_queue: [],
    approvals: [],
    performance: {},
    alerts: [],
    primary_action: {
      label: "unknown",
      action: "unknown",
      context_id: "unknown",
    },
    system_health: "unknown",
  };
}

function getSupabaseConfig() {
  const url = (process.env.SUPABASE_URL || process.env.NEXT_PUBLIC_SUPABASE_URL || "").replace(/\/$/, "");
  const key =
    process.env.SUPABASE_SERVICE_ROLE_KEY ||
    process.env.SUPABASE_ANON_KEY ||
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY ||
    "";
  return { url, key };
}

async function supabaseRequest(path, options) {
  const { url, key } = getSupabaseConfig();
  if (!url || !key) return null;
  try {
    const resp = await fetch(`${url}/rest/v1/${path}`, {
      ...options,
      headers: {
        apikey: key,
        Authorization: `Bearer ${key}`,
        "Content-Type": "application/json",
        Prefer: options?.prefer || "return=representation",
        ...(options?.headers || {}),
      },
    });
    if (!resp.ok) return null;
    return await resp.json();
  } catch {
    return null;
  }
}

function asList(value) {
  return Array.isArray(value) ? value : [];
}

function asDict(value) {
  return value && typeof value === "object" && !Array.isArray(value) ? value : {};
}

function asString(value, fallback) {
  if (value == null) return fallback || "unknown";
  const text = String(value).trim();
  return text || fallback || "unknown";
}

function asNumber(value) {
  const n = parseFloat(value);
  return Number.isFinite(n) ? n : 0;
}

async function fetchTrends() {
  const rows = await supabaseRequest(
    `${DEFAULT_FEED_TABLE}?source=eq.tiktok&select=timestamp,type,signal_strength,virality_score,trend_state,summary,dedupe_key&order=timestamp.desc&limit=50`,
    { method: "GET" }
  );
  if (!Array.isArray(rows)) return [];
  return rows.map((row) => ({
    id: asString(row.dedupe_key || row.summary),
    type: asString(row.type),
    signal_strength: row.signal_strength || 0,
    virality_score: row.virality_score || 0,
    trend_state: asString(row.trend_state),
    summary: asString(row.summary),
    timestamp: asString(row.timestamp),
  }));
}

async function fetchCachedProducts(accountId, key) {
  let path = `${DEFAULT_CACHE_TABLE}?select=payload&order=updated_at.desc&limit=1`;
  if (accountId) path += `&account_id=eq.${encodeURIComponent(accountId)}`;
  const rows = await supabaseRequest(path, { method: "GET" });
  if (!Array.isArray(rows) || !rows.length) return [];
  let payload = rows[0].payload || {};
  if (typeof payload === "string") {
    try {
      payload = JSON.parse(payload);
    } catch {
      payload = {};
    }
  }
  const products = asList(payload[key]);
  return products
    .filter((item) => item && typeof item === "object")
    .map((item) => ({
      name: asString(item.product || item.name),
      signal_strength: item.signal_strength || item.score || 0,
      source: key === "emerging_products" ? "emerging" : "trending",
      confidence: item.confidence || 0,
      evidence: asList(item.evidence),
    }));
}

function mergeProducts(emerging, trending) {
  const merged = [];
  const seen = new Set();
  for (const item of [...emerging, ...trending]) {
    const key = asString(item.name, "").toLowerCase();
    if (!key || seen.has(key)) continue;
    seen.add(key);
    merged.push(item);
  }
  return merged;
}

async function fetchContentQueue(accountId) {
  if (!accountId) return [];
  const rows = await supabaseRequest(
    `${DEFAULT_QUEUE_TABLE}?account_id=eq.${encodeURIComponent(accountId)}&select=id,account_id,caption,hashtags,hook,product_name,status,scheduled_time,created_at&order=created_at.desc&limit=50`,
    { method: "GET" }
  );
  if (!Array.isArray(rows)) return [];
  return rows.map((row) => ({
    id: asString(row.id, ""),
    account_id: asString(row.account_id, ""),
    caption: asString(row.caption),
    hashtags: asList(row.hashtags),
    hook: asString(row.hook),
    product_name: asString(row.product_name),
    status: asString(row.status),
    scheduled_time: asString(row.scheduled_time),
    created_at: asString(row.created_at),
  }));
}

async function fetchApprovals(accountId) {
  if (!accountId) return [];
  const rows = await supabaseRequest(
    `${DEFAULT_QUEUE_TABLE}?account_id=eq.${encodeURIComponent(accountId)}&status=in.(pending,queued)&select=id,caption,product_name,status,created_at&order=created_at.desc&limit=25`,
    { method: "GET" }
  );
  if (!Array.isArray(rows)) return [];
  return rows.map((row) => ({
    content_id: asString(row.id, ""),
    caption: asString(row.caption),
    product_name: asString(row.product_name),
    status: asString(row.status),
    created_at: asString(row.created_at),
  }));
}

async function fetchPerformance(accountId) {
  if (!accountId) return {};
  const rows = await supabaseRequest(
    `${DEFAULT_PERFORMANCE_TABLE}?account_id=eq.${encodeURIComponent(accountId)}&select=content_id,performance_metrics,timestamp&order=timestamp.desc&limit=25`,
    { method: "GET" }
  );
  if (!Array.isArray(rows)) return {};
  let totalViews = 0;
  let totalEngagement = 0;
  const snapshots = rows.map((row) => {
    const metrics = asDict(row.performance_metrics);
    totalViews += asNumber(metrics.views);
    totalEngagement += asNumber(metrics.engagement_rate);
    return {
      content_id: asString(row.content_id, ""),
      metrics,
      timestamp: asString(row.timestamp),
    };
  });
  const count = snapshots.length;
  return {
    snapshot_count: count,
    total_views: totalViews,
    avg_engagement_rate: count ? Math.round((totalEngagement / count) * 10000) / 10000 : 0,
    snapshots,
  };
}

async function computeSystemHealth() {
  const rows = await supabaseRequest(`${DEFAULT_QUEUE_TABLE}?select=id&limit=1`, { method: "GET" });
  if (!Array.isArray(rows)) return "failing";
  return "degraded";
}

function buildAlerts(inventoryGaps, systemHealth) {
  const alerts = [];
  for (const gap of inventoryGaps) {
    alerts.push({
      level: "warning",
      code: "inventory_gap",
      message: `Inventory gap: ${asString(gap.product_name)}`,
    });
  }
  if (systemHealth === "failing") {
    alerts.push({ level: "error", code: "system_health", message: "System health is failing" });
  } else if (systemHealth === "degraded") {
    alerts.push({ level: "warning", code: "system_health", message: "System health is degraded" });
  }
  return alerts;
}

function deriveFlowAndAction(trends, products, contentQueue, approvals, inventoryGaps, systemHealth) {
  const todayFlow = {
    step: "trend → product → content → queue",
    next_action: "unknown",
    status: "unknown",
  };
  const primaryAction = { label: "unknown", action: "unknown", context_id: "unknown" };

  if (inventoryGaps.length) {
    todayFlow.next_action = "Resolve inventory gaps before attaching products";
    todayFlow.status = "blocked";
    primaryAction.label = "Fix inventory gap";
    primaryAction.action = "resolve_inventory_gap";
    primaryAction.context_id = asString(inventoryGaps[0].product_name);
    return { todayFlow, primaryAction };
  }
  if (approvals.length) {
    todayFlow.next_action = "Review pending content approvals";
    todayFlow.status = "awaiting_approval";
    primaryAction.label = "Approve content";
    primaryAction.action = "approve_content";
    primaryAction.context_id = asString(approvals[0].content_id);
    return { todayFlow, primaryAction };
  }
  if (!contentQueue.length && products.length) {
    todayFlow.next_action = "Generate content from detected products";
    todayFlow.status = "ready_for_content";
    primaryAction.label = "Generate content";
    primaryAction.action = "generate_content";
    primaryAction.context_id = asString(products[0].name);
    return { todayFlow, primaryAction };
  }
  if (!products.length && trends.length) {
    todayFlow.next_action = "Match products to active trends";
    todayFlow.status = "trend_detected";
    primaryAction.label = "Match products";
    primaryAction.action = "match_products";
    primaryAction.context_id = asString(trends[0].id || trends[0].summary);
    return { todayFlow, primaryAction };
  }
  if (contentQueue.length) {
    todayFlow.next_action = "Monitor queued content pipeline";
    todayFlow.status = "in_queue";
    primaryAction.label = "View queue";
    primaryAction.action = "view_queue";
    primaryAction.context_id = asString(contentQueue[0].id);
    return { todayFlow, primaryAction };
  }
  todayFlow.status = systemHealth;
  todayFlow.next_action = "Run trend scan to refresh signals";
  primaryAction.label = "Refresh trends";
  primaryAction.action = "run_trend_scan";
  primaryAction.context_id = "unknown";
  return { todayFlow, primaryAction };
}

async function assembleLiveState(accountId) {
  const state = emptyContract();
  const account = asString(accountId, "").trim();

  try {
    const [trends, emerging, trending, contentQueue, approvals, performance, systemHealth] =
      await Promise.all([
        fetchTrends(),
        fetchCachedProducts(account, "emerging_products"),
        fetchCachedProducts(account, "trending_products"),
        fetchContentQueue(account),
        fetchApprovals(account),
        fetchPerformance(account),
        computeSystemHealth(),
      ]);

    const products = mergeProducts(emerging, trending);
    const inventoryGaps = [];
    const inventoryPrevention = [];
    const { todayFlow, primaryAction } = deriveFlowAndAction(
      trends,
      products,
      contentQueue,
      approvals,
      inventoryGaps,
      systemHealth
    );

    state.today_flow = todayFlow;
    state.trends = trends;
    state.products = products;
    state.inventory_gaps = inventoryGaps;
    state.inventory_prevention = inventoryPrevention;
    state.content_queue = contentQueue;
    state.approvals = approvals;
    state.performance = performance;
    state.alerts = buildAlerts(inventoryGaps, systemHealth);
    state.primary_action = primaryAction;
    state.system_health = systemHealth;
    return state;
  } catch {
    return state;
  }
}

module.exports = { emptyContract, assembleLiveState };
